dagql/idtui/db.go

```diff
@@ -0,0 +1,257 @@
+	for _, t := range status.Tasks {
+		db.recordTask(t)
+	}
+
+	// track parent/child vertices
+	for _, v := range status.Children {
+		if db.Children[v.Vertex] == nil {
+			db.Children[v.Vertex] = make(map[string]struct{})
+		}
+		for _, out := range v.Vertexes {
+			db.Children[v.Vertex][out] = struct{}{}
+		}
+	}
+
+	return nil
+}
+
+func (db *DB) recordTask(t *progrock.VertexTask) {
+	tasks := db.Tasks[t.Vertex]
+	var updated bool
+	for i, task := range tasks {
+		if task.Name == t.Name {
+			tasks[i] = t
+			updated = true
+		}
+	}
+	if !updated {
+		tasks = append(tasks, t)
+		db.Tasks[t.Vertex] = tasks
+	}
+}
+
+// Step returns a Step for the given digest if and only if the step should be
+// displayed.
+//
+// Currently this means:
+//
+// - We don't show `id` selections, since that would be way too much noise.
+// - We don't show internal non-ID vertices, since they're not interesting.
+// - We DO show internal ID vertices, since they're currently marked internal
+// just to hide them from the old TUI.
+func (db *DB) Step(dig string) (*Step, bool) {
+	step := &Step{
+		Digest: dig,
+		db:     db,
+	}
+
+	outID := db.IDs[dig]
+	switch {
+	case outID != nil && outID.Field == "id":
+		return nil, false
+	case outID == nil && step.IsInternal():
+		return nil, false
+	case outID != nil && !step.HasStarted():
+		return nil, false
+	case outID == nil && !step.HasStarted():
+		return nil, false
+	}
+	if outID != nil && outID.Base != nil {
+		parentDig, err := outID.Base.Digest()
+		if err != nil {
+			return nil, false
+		}
+		step.Base = db.Simplify(parentDig.String())
+	}
+	return step, true
+}
+
+func (db *DB) HighLevelStep(id *idproto.ID) (*Step, bool) {
+	parentDig, err := id.Digest()
+	if err != nil {
+		return nil, false
+	}
+	return db.Step(db.Simplify(parentDig.String()))
+}
+
+func (db *DB) PrimaryVertex(dig string) *progrock.Vertex {
+	var earliest *progrock.Vertex
+	vs := make([]*progrock.Vertex, 0, len(db.Intervals[dig]))
+	for _, vtx := range db.Intervals[dig] {
+		vs = append(vs, vtx)
+	}
+	sort.Slice(vs, func(i, j int) bool {
+		return vs[i].Started.AsTime().Before(vs[j].Started.AsTime())
+	})
+	for _, vtx := range db.Intervals[dig] {
+		if vtx.Cached {
+			continue
+		}
+		if earliest == nil {
+			earliest = vtx
+			continue
+		}
+		// if vtx.Completed == nil && earliest.Completed != nil {
+		// 	// prioritize actively running vertex over a completed one
+		// 	earliest = vtx
+		// 	continue
+		// }
+		// if earliest.Completed == nil && vtx.Completed != nil {
+		// 	// never replace a running vertex with a completed one
+		// 	continue
+		// }
+		if vtx.Started.AsTime().Before(earliest.Started.AsTime()) {
+			earliest = vtx
+		}
+	}
+	return earliest
+}
+
+func (db *DB) IsTransitiveDependency(dig, depDig string) bool {
+	for _, v := range db.Intervals[dig] {
+		for _, dig := range v.Inputs {
+			if dig == depDig {
+				return true
+			}
+			if db.IsTransitiveDependency(dig, depDig) {
+				return true
+			}
```

Comment:
This looks like a DAG walk that doesn't use a memo to de-dupe, right? If so, doing that in various forms has been the source of at least 3 bugs in the past between buildkit+dagger where everything is fine until `n` exceeds some threshold and everything just hangs suddenly. Not sure if in the cards here, just checking.

Reply:
:thinking: if this doesn't terminate, then the DAG would have cycles right?

Reply:
Good point, we don't actually need this for the TUI (it's a holdover from Dagviz) so I'll just comment it out.